File: backend/models/stock.go
package models

/*
	Models define the structure of the data used in the application,
	such as Stock and ApiResponse.
*/

import "time"

// StockRatings represents a stock rating entry.
type StockRatings struct {
	ID         int       `json:"id" db:"id" example:"1"`
	Ticker     string    `json:"ticker" db:"ticker" example:"AAPL"`
	TargetFrom string    `json:"target_from" db:"target_from" example:"$150.00"`
	TargetTo   string    `json:"target_to" db:"target_to" example:"$180.00"`
	Company    string    `json:"company" db:"company" example:"Apple Inc."`
	Action     string    `json:"action" db:"action" example:"target raised by"`
	Brokerage  string    `json:"brokerage" db:"brokerage" example:"Goldman Sachs"`
	RatingFrom string    `json:"rating_from" db:"rating_from" example:"Buy"`
	RatingTo   string    `json:"rating_to" db:"rating_to" example:"Strong Buy"`
	Time       time.Time `json:"time" db:"time" example:"2025-01-15T10:30:00Z"`
	CreatedAt  time.Time `json:"created_at" db:"created_at" example:"2025-01-15T10:35:00Z"`
}

// ApiResponse represents the response from the external stock API.
type ApiResponse struct {
	Items    []StockRatings `json:"items"`
	NextPage string         `json:"next_page"`
}

// PageRequest represents the expected structure of the pagination request.
type PageRequest struct {
	Page int `json:"page" binding:"required" example:"1"`
}

type BulkPageRequest struct {
	StartPage int `json:"start_page" binding:"required" example:"1"`
	EndPage   int `json:"end_page" binding:"required" example:"100"`
}

type PaginationRequest struct {
	PageNumber int `json:"page_number" binding:"required" example:"1"`
	PageLength int `json:"page_length" binding:"required" example:"20"`
}

File: frontend/src/types/stock.ts
export interface Stock {
  ticker: string;
  company: string;
  brokerage: string;
  action: string;
  rating_from: string;
  rating_to: string;
  target_from: string;
  target_to: string;
  time: string;
}

export interface ApiResponse {
  items: Stock[];
  next_page: string;
}

export type SortField = 'ticker' | 'company' | 'brokerage' | 'target_from' | 'target_to' | 'action';
export type SortDirection = 'asc' | 'desc';

export interface StockFilters {
  search: string;
  action: string;
  rating_from: string;
  rating_to: string;
  target_from_min: number;
  target_from_max: number;
  target_to_min: number;
  target_to_max: number;
}

export interface FilterOptions {
  actions: string[];
  ratings_from: string[];
  ratings_to: string[];
  target_ranges: {
    min_price: number;
    max_price: number;
  };
}